R/constructors-Node.R

```diff
@@ -133,25 +133,44 @@ setMethod("Node", "missing", function(id, x = numeric(), y = numeric(), color =
 #' 
 #' Generate a NodeList from an edgeList
 #' @param object Object containing data to be converted to a NodeList
+#' @param layout string indicating the layout algorithm to be used. Options are 'none', 'force', 
+#' 'circle' or 'nicely' which are implemented via igraph. Defaults to 'nicely'. 
 #' @return NodeList
 #' @export
 #' @examples
 #' NodeList(data.frame(source='a',target='b'))
-setGeneric("NodeList", function(object) standardGeneric("NodeList"))
+setGeneric("NodeList", function(object, ...) standardGeneric("NodeList"))
 
 #' @export
-setMethod("NodeList", "data.frame", function(object = data.frame(source=character(),target=character())) {
+setMethod("NodeList", "data.frame", function(object = data.frame(source=character(),target=character()), layout = c("nicely", "force", "circle", "none")) {
   if (!inherits(isValidEdgeList(object), "logical")) {
     stop(paste("Invalid edgeList:", isValidEdgeList(object), collapse = '\n'))
   }
+  layout <- veupathUtils::matchArg(layout)
   
-  allNodeIds <- c(object$source, object$target)
+  graph <- igraph::graph_from_data_frame(object, directed = FALSE)
+  if (layout != "none") {
+    if (layout == "force") {
+      coords <- igraph::layout_with_fr(graph)
+    } else if (layout == "circle") {
+      coords <- igraph::layout_in_circle(graph)
+    } else if (layout == "nicely") {
+      coords <- igraph::layout_nicely(graph)
+    } else {
+      stop("layout must be 'force', 'circle' or 'nicely'")
+    }
+    rownames(coords) <- names(igraph::V(graph))
+  }
 
-  makeNodeWithDegree <- function(nodeId, allNodeIds) {
-    new("Node", id = NodeId(nodeId), degree = length(which(allNodeIds == nodeId)))
+  # if we want to move this out of the constructor it needs to have graph and coords passed to it
```

Comment:
agreed we should move it out of constructors. This is a thing dan would want to do

Reply:
i can move it i suppose, but it seems we just as easily could later if an actual use case came up. in either case, i dont understand what it is dan would be wanting here.

Reply:
I guess if we leave dan to use igraph then this can stay as is. I keep thinking general purpose for how this should all work but it's just the ds for now that needs to do this.

Reply:
i guess at the heart of it, the point is that being able to get a specific node w the degree and coords populated out of an igraph object is generally useful? thats interesting, i hadnt anticipated that. but i also think how to do that best for general audiences is a thing we can save for later.